File: packages/flutter/lib/src/logic/pagination.dart
/// 分页规则的 Dart 移植。
///
/// Flutter 无法复用 TypeScript，因此这里按同一套算法重写；
/// scripts/check-parity.mjs 会用同一组用例校验两边输出一致，防止分叉。

/// 页码序列里的一项：要么是具体页码，要么是一处省略
class IPageItem {
  const IPageItem.page(this.page) : gap = null;
  const IPageItem.gap(this.gap) : page = null;

  final int? page;

  /// 'left' 或 'right'
  final String? gap;

  bool get isPage => page != null;
  bool get isLeftGap => gap == 'left';
}

int pageCountOf(int total, int pageSize) {
  final size = pageSize < 1 ? 1 : pageSize;
  final count = (total / size).ceil();
  return count < 1 ? 1 : count;
}

int clampPage(int page, int pageCount) {
  final max = pageCount < 1 ? 1 : pageCount;
  if (page < 1) return 1;
  return page > max ? max : page;
}

/// 始终保留首尾页，中间窗口跟随当前页滑动，断开处以省略占位
List<IPageItem> buildPages(int current, int pageCount, [int maxVisible = 5]) {
  final count = pageCount < 1 ? 1 : pageCount;
  final window = maxVisible < 1 ? 1 : maxVisible;

  if (count <= window + 2) {
    return List.generate(count, (i) => IPageItem.page(i + 1));
  }

  final half = window ~/ 2;
  var start = current - half;
  if (start < 2) start = 2;
  var end = start + window - 1;
  if (end >= count) {
    end = count - 1;
    start = end - window + 1;
    if (start < 2) start = 2;
  }

  final result = <IPageItem>[const IPageItem.page(1)];
  if (start > 2) result.add(const IPageItem.gap('left'));
  for (var i = start; i <= end; i++) {
    result.add(IPageItem.page(i));
  }
  if (end < count - 1) result.add(const IPageItem.gap('right'));
  result.add(IPageItem.page(count));
  return result;
}

String rangeText(int current, int pageSize, int total) {
  if (total == 0) return '共 0 条';
  final from = (current - 1) * pageSize + 1;
  final rawTo = current * pageSize;
  final to = rawTo > total ? total : rawTo;
  return '第 $from-$to 条 / 共 $total 条';
}

File: packages/common/src/logic/pagination.ts
/**
 * 页码序列计算：始终保留首尾页，中间窗口跟随当前页滑动，
 * 断开处以 'left' / 'right' 占位。各端只负责把它渲染成按钮。
 */
export type PageItem = number | 'left' | 'right'

export function buildPages(
  current: number,
  pageCount: number,
  maxVisible = 5
): PageItem[] {
  const count = Math.max(1, pageCount)
  const window = Math.max(1, maxVisible)
  if (count <= window + 2) return Array.from({ length: count }, (_, i) => i + 1)

  const half = Math.floor(window / 2)
  let start = Math.max(2, current - half)
  let end = start + window - 1
  if (end >= count) {
    end = count - 1
    start = Math.max(2, end - window + 1)
  }

  const result: PageItem[] = [1]
  if (start > 2) result.push('left')
  for (let i = start; i <= end; i++) result.push(i)
  if (end < count - 1) result.push('right')
  result.push(count)
  return result
}

export function pageCountOf(total: number, pageSize: number) {
  return Math.max(1, Math.ceil(total / Math.max(1, pageSize)))
}

export function clampPage(page: number, pageCount: number) {
  return Math.min(Math.max(1, page), Math.max(1, pageCount))
}

export function rangeText(current: number, pageSize: number, total: number) {
  if (!total) return '共 0 条'
  const from = (current - 1) * pageSize + 1
  const to = Math.min(current * pageSize, total)
  return `第 ${from}-${to} 条 / 共 ${total} 条`
}
